Python/chapter01/1.7 - Rotate Matrix/miguel_1.7_sol.py

```diff
@@ -0,0 +1,185 @@
+"""
+Python version 3.7.0
+1.7 - Rotate Matrix
+Given an image represented by an NxN matrix, where each pixel in the image is 4 bytes, write a method to rotate
+the image by 90 degrees.  Can you do this in place?
+"""
+import unittest
+from typing import List, Optional
+
+
+def rotate_matrix(matrix: List[List[int]]) -> List[List[int]]:
+    """
+    Rotate matrix will rotate the given matrix by 90 degrees.
+    Runtime: O(N^2), asymptotic runtime depends on N
+    Space Complexity: O(N^2), creating a new matrix of NxN called 'rotated'
+    :param matrix: an NxN matrix
+    :param N: the size of the matrix (NxN)
+    :return: a newly rotated matrix
+    """
+    N = len(matrix)
+    rotated = [[0 for i in range(N)] for j in range(N)]
+    for i in range(N):
+        for j in range(N):
+            rotated[j][(N-1)-i] = matrix[i][j]
+    return rotated
+
+
+def _perform_full_rotation(matrix: List[List[int]], row: int, col: int, N: Optional[int] = None):
+    """
+    Helper function that performs four 90 degree rotations starting at row, col.
+    :param matrix: an NxN matrix
+    :param row: starting row
+    :param col: starting column
+    :param N: size of matrix
+    :return:
+    """
+    if N is None:
+        N = len(matrix)
+    num_rotations = 4
+    start_row = row
+    rotated_row = row
+    rotated_col = col
+    temp_new = matrix[row][col]
+    for r in range(0, num_rotations):
+        temp = temp_new
+        # compute new rotated indices
+        prev_col = rotated_col
+        rotated_col = N - 1 - rotated_row + (start_row * 2)  # offset to account for reduced N
+        rotated_row = prev_col
+        # store value at newly computed indices
+        temp_new = matrix[rotated_row][rotated_col]
+        matrix[rotated_row][rotated_col] = temp
+
+
+def rotate_matrix_in_place(
+        matrix: List[List[int]],
+        start_row: int = 0,
+        start_col: int = 0,
+        N: Optional[int] = None) -> List[List[int]]:
+    """
+    Does the same as rotate_matrix, but in place.
+    Runtime: O(N^2), asymptotic runtime depends on N. We make N^2 swaps.
+    Space Complexity: O(1), constant amount of temp variables that does not depend on N.
+                        Note: I am assuming that compiler will optimize the tail recursion.
+    :param matrix: an NxN matrix
+    :param start_row: starting row index
+    :param start_col: starting col index
+    :param N: the size of the matrix (NxN)
+    :return: the input matrix, but rotated
+    """
+
+    if N is None:
+        N = len(matrix)
+    if N == 0 or N == 1:
```

Comment:
what you are doing in this method is initializing:

```
start_row = 0
start_col = 0 # this is identical to start_row, so you don't really need it
N = len(matrix)
```

then repeatedly do:

```python
_perform_full_rotation(matrix, start_row, start_col, N)
start_row += 1
start_col += 1
N -= 2
```

until `N < 2`. That looks and feels like a for loop. furthermore, you can derive those three variables from the loop variable.

Reply:
The for loop you are suggesting would get rid of the recursive calls, right?  Do you recommend doing it that way instead of the recursive approach?  If so, why?

Reply:
that's correct. loops are significantly easier to understand than recursion (in most cases). the fact that the loop uses O(1) memory is an added bonus.